### PR TITLE
Stream Directories and Files using Flyte

### PR DESCRIPTION
**Pull request by kumare3**
# TL;DR

This PR implements streaming FlyteFile and allows walking FlyteDirectory.
the streaming FlyteFile returns a Filelike object that is supported by fsspec.

**Note** this is still experimental and we would love feedback on the API.

*Example: Accept a flytefile and stream to another FlyteFile, using simple caching*
```python
@task
def copy_file(ff: FlyteFile) -> FlyteFile:
    new_file = FlyteFile.new_remote_file(ff.remote_path)
    with ff.open("r", cache_type="simplecache", cache_options={}) as r:
        with new_file.open("w") as w:
            w.write(r.read())
    return new_file
```

*Example: Accept a flyteDirectory and walk it, while copying the directory file by file to another Flyte directory*
```python
@task
def process_folder(fd: FlyteDirectory) -> FlyteDirectory:
    out_fd = FlyteDirectory.new_remote()
    for base, x in fd.crawl():
        src = os.path.join(base, x)
        out_file = out_fd.new_file(x)
        with FlyteFile(src).open("rb") as f:
            with out_file.open("wb") as o:
                o.write(f.read())
    return out_fd
```

*Alternatively, Files can always be handled by fsspec streaming*
```python
            @task
            def copy_file(ff: FlyteFile) -> FlyteFile:
                new_file = FlyteFile.new_remote_file(ff.name)
                fs = flytekit.get_filesystem()
                with fs.open(f"readahead::{ff.remote_path}", "rb", readahead={}) as r:
                    with new_file.open("wb") as w:
                        w.write(r.read())
                return new_file
```

## Type
 - [ ] Bug Fix
 - [x] Feature
 - [ ] Plugin

## Are all requirements met?

 - [x] Code completed
 - [x] Smoke tested
 - [x] Unit tests added
 - [x] Code documentation added
 - [x] Any pending items have an associated Issue

## Complete description

Other minor changes:
* Get rid of sqlite data download dependency by adding it to this repo
* Make the base to_html function not abstract.

## Tracking Issue
https://github.com/flyteorg/flyte/issues/3334

### CODE REVIEW

**Comment by kumare3** [comment]
The current test failures are because the requirement of fsspec is not met. We can add it only for `test` dependencies as this is runtime, or better we can wait for the subsystem re-write coming soon

**Review by eapolinario — COMMENTED** [review]
What sort of tests do we need?

File: flytekit/types/file/file.py [hunk 1]
```diff
@@ -226,6 +236,62 @@ def remote_source(self) -> str:
...
+        try:
+            final_path = self.path
+            if self.remote_source:
+                final_path = self.remote_source
+            elif self.remote_path:
+                final_path = self.remote_path
+            fs = ctx.file_access.get_filesystem_for_path(final_path, cache_type=cache_type, cache_options=cache_options)
+            yield fs.open(final_path, mode)
+        except ImportError as e:
+            print(
```
[the remark this answers is not in the document]
Comment (wild-endeavor) [reply flytekit/types/file/file.py:289]: removing try/catch since we're now requiring it

File: tests/flytekit/unit/types/directory/test_types.py [hunk 2]
```diff
@@ -0,0 +1,38 @@
+from pathlib import Path
+
+from flytekit import FlyteContext
+from flytekit.types.directory import FlyteDirectory
+from flytekit.types.file import FlyteFile
+
+
+def test_crawl(tmpdir):
```
[the remark this answers is not in the document]
Comment (wild-endeavor) [reply tests/flytekit/unit/types/directory/test_types.py:8]: deleting whole test, added another one

File: flytekit/types/directory/types.py [hunk 3]
```diff
@@ -148,6 +152,11 @@ def __fspath__(self):
     def extension(cls) -> str:
         return ""
 
+    @classmethod
+    def new_remote(cls) -> FlyteDirectory:
```
[the remark this answers is not in the document]
Comment (wild-endeavor) [reply flytekit/types/directory/types.py:156]: what's a docstring

**Review by wild-endeavor — APPROVED** [review]

**Review by pingsutw — APPROVED** [review]
